### ISSUE TITLE
Proposal for a better version of the select operator

### ISSUE THREAD

**Issue by BioPhoton** [rx-angular/rx-angular#263]
# Problem solved by the feature

ATM we have 2 main operators/ways to shortcut state selections. `select` and `selectSlice`.

## `select` 
`select` is a standalone operator as well as a method on `RxState`. It plucks a (deeply) nested property or takes a series of operators the derive state.

A string-based version that takes a series of keys, reflecting a deeply nested value in an object: 
```typescript
 select<K1 extends keyof T, K2 extends keyof T[K1]>(
    k1: K1,
    k2: K2,
   // more property names
  ): Observable<T[K1][K2][...]>;
```
And a reactive version that takes a series of functions that adhere to the signature of an RxJS operator (`(Observable<T>) => Observable<T|O>`)

```typescript
select<A = T, B = A>(
    op1: OperatorFunction<T, A>,
    op2: OperatorFunction<A, B>,
    // more operators
  ): Observable<D>;
```

Pros:
- Both versions turn the stream into a "stateful stream", a stream that emits only distinct values and replays the last one on subscribe.
- The string version is super nice to pluck a single property from the state and get a single distinct replayed observable of it.
- The operator version can compose any RxJS operator to derive state and apply behavior. 

Cons:
- The string version does not allow to select a subset of the shallow properties of an object. (only nested properties)
- No way to finetune distinct values

## `selectSlice` 
`selectSlice` is a standalone operator that takes a set of properties and an optional set of compare functions to select a state slice. 

The shallow property names:

```typescript
selectSlice<T extends object, K extends keyof T>(
  keys: K[]
): OperatorFunction<T, PickSlice<T, K>> 
```

And an optional compare map:

```typescript
selectSlice<T extends object, K extends keyof T>(
  keys: K[],
  keyCompareMap: KeyCompareMap<{ [P in K]: T[P] }>
): OperatorFunction<T, PickSlice<T, K>> 
```

Pros:
- Perfect for deriving a slice of the state (a subset of the shallow properties).
- Easy to tweak what distinct values from another. (`KeyCompareMap`)

Cons:
- The resulting stream is not "stateful"
- No options to pass operators

The problem here is that `selectSlice` is often used within select like this:

```typescript
state.select(
    selectSlice(['prop1', 'prop2', 'prop3']),
    map(s => ({
        ...s,
        num: Object.keys(prop2).length
    })),
    // optionally used with transformationHelpers
    // map(s => setProp( s, 'num', Object.keys(s.prop2).length))
)
```

As the main features of the 2 operators are so different you end up often using both. 
Combining the best features of both in a way we don't break too many people could bring a lot of benefits.

# Solution

TL;DR
> Ditch deep pluck over shallow and distinction functions over operator composition.

As we have to break as little people as possible we need a good migration strategy to introduce the change.

I make following assumption: Very view people pluck deeply nested properties. 😇 (please correct me)
Based on that assumption I would deprecate the other optional strings. 

The first parameter currently can be a string or a function.
I would change that to be a string a function or an array of strings and keep the rest of the params as operators.

- The user will either only put operators there, or plucks shallow subset of the properties. 
- If a single property is needed we can put just a string there.
- As an equivalent to `selectSlice`, the operator can take multiple operators.
- The resulting observable is "stateful" same as the original `select`.

## Proposal

```typescript
selectSlice<T extends object, K extends keyof T, O>(
  opOrKeyOrKeys: OperatorFunction<T, A>, K, K[],
  op1: OperatorFunction<A, B>,
  op2: OperatorFunction<B, O>,
  // more operators
): OperatorFunction<T, PickSlice<T, K> | O>
```

The param handling could look like this:
```typescript
selectSlice<T extends object, K extends keyof T, O>(
  opOrKeyOrKeys: OperatorFunction<T, A>, K, K[], 
  ...opsOrStrings: Function [] | string[]
  // more operators
): OperatorFunction<T, PickSlice<T, K> | O>
```

As a comparison, the above state selection with selectSlice would look like this;

```typescript
state.select(['prop1', 'prop2', 'prop3'],
    map(s => setProp( s, 'num', Object.keys(s.prop2).length))
)
```

Versions:

- Single string
```typescript
state.select('prop1')
```

- Single operator
```typescript
state.select(map( s => s.prop1))
```

- Multiple operators
```typescript
state.select(
    map( s => s.prop1),
    debounce(350),
    // ...
)
```

- Shallow slice:
```typescript
state.select(['prop1', 'prop2', 'prop3'])
```

- Single string and operators
```typescript
state.select('prop1',
    map( s => s.prop1),
    debounce(350),
    // ...
)
```

- Shallow properties array and operators
```typescript
state.select['prop1', 'prop2', 'prop3'],
    map( s => s.prop1),
    debounce(350),
    // ...
)
```

## Migration path

1. Deprecate all but the first string parameter
2. Introduce an optional k[] for the first parameter
2.1. Do the same for the standalone operator 
3. As we remove deep pluck we have to expose the internal safePluck to the public API to provide a migration path for that usage. 
4. Break (remove) the string properties. 

# Alternatives considered

- Adding another method on the RxState class.
  Adding another method of the state class would confuse which one to use.
  It would weaken the simplicity of the state interface. 2 imperative and 2 reactive methods.

- Keeping the deep pluck from select
  too many overloads
  more complex to understand

# Additional Context

n/a

**Comment by hoebbelsB** [comment]
Thank you very much! This is way better than what I had in mind.

> I make following assumption: Very view people pluck deeply nested properties. innocent (please correct me)

I think this is true. Accessing deeply nested properties is more of an edge case scenario than every day usage. At least when it comes to local state.

Currently I cannot think about any other version than the four you already explained. Or is the `KeyCompareMap` still useful as an overload?

> Adding another method on the RxState class.
Adding another method of the state class would confuse which one to use.
It would weaken the simplicity of the state interface. 2 imperative and 2 reactive methods.

I totally agree.

> As an equivalent to selectSlice, the operator can take multiple operators.

What does this mean exactly? I thought it is not about an operator, it is about the `select` method of the state?

> Deprecate all but the first string parameter
    Introduce an optional k[] for the first parameter
    Break (remove) the string properties.

How does this look like? We will need to do major version upgrade, don't we?

**Comment by BioPhoton** [comment]
> What does this mean exactly? I thought it is not about an operator, it is about the select method of the state?

Yes and no :) it is about the `RxState#select` method and the standalone `select` operator.

I also updated the comment with answers to your other questions.

**Comment by matejchalk** [comment]
I was looking into this issue for a little while and had some concerns with the proposed API changes to `select`:

1. It looks like the key-comparison map would be dropped (the 2nd parameter of `selectSlice`). Is this feature not being used? It seems pretty useful.
2. When combining the single key (from current `select`) with array of keys (from current `selectSlice`) - the problem is the first currently does a "pluck" (you get the property value directly, without the object), while the other does a "pick" (you get the object without unspecified properties). Combining these two ways in one would be pretty confusing, I think (e.g. `select('key')` would be different from `select(['key'])`). Could be reconciled by doing a "pick" in both cases, but that would break the existing `select('key')` use-cases (you'd have to add a `map`).

I wonder if it wouldn't be a better solution to keep `select` and `selectSlice` separate and make these changes instead:
1. Modify the `select` overloads to support `select('key', operator1, operator2, ...)` (optionally also drop support for `select('key1', 'key2', ...)` if that makes it easier).
2. Use `stateful` within `selectSlice` - this might actually mean no API changes, not sure.

**Comment by hoebbelsB** [comment]
@matejchalk i actually like that idea pretty much. would you like to PR this?

**Comment by matejchalk** [comment]
> @matejchalk i actually like that idea pretty much. would you like to PR this?

Yeah, sure :slightly_smiling_face:

**Comment by BioPhoton** [comment]
Still I see the shell copy a more useful thing then deep pluck in the top level api of state...

Maybe something for v2?

**Comment by hoebbelsB** [comment]
won't that be fixed by: (the optional stuff)?

> * Modify the `select` overloads to support `select('key', operator1, operator2, ...)` (optionally also drop support for `select('key1', 'key2', ...)` if that makes it easier).

Then there is no deep pluck anymore.

Additionally the `selectSlice` method would take a set of `OperatorFunctions` as well, just like the `select` method?

`selectSlice(['key', 'key2'], operator1, operator2, ...)`
